File: lmkkmeans_train.m
function state = lmkkmeans_train(Km, parameters)
    tic;
    N = size(Km, 2);
    P = size(Km, 3);
    Theta = ones(N, P) / P;
    K_Theta = calculate_localized_kernel_theta(Km, Theta);

    opt.disp = 0;
    objective = zeros(parameters.iteration_count, 1);
    for iter = 1:parameters.iteration_count
        fprintf(1, 'running iteration %d...\n', iter);
        [H, ~] = eigs(K_Theta, parameters.cluster_count, 'la', opt);
        HHT = H * H';
        
        Q = zeros(N * P, N * P);
        for m = 1:P
            start_index = (m - 1) * N + 1;
            end_index = m * N;
            Q(start_index:end_index, start_index:end_index) = eye(N, N) .* Km(:, :, m) - HHT .* Km(:, :, m);
        end
        res = mskqpopt(Q, zeros(N * P, 1), repmat(eye(N, N), 1, P), ones(N, 1), ones(N, 1), zeros(N * P, 1), ones(N * P, 1), [], 'minimize echo(0)');
        Theta = reshape(res.sol.itr.xx, N, P);
        K_Theta = calculate_localized_kernel_theta(Km, Theta);

        objective(iter) = trace(H' * K_Theta * H) - trace(K_Theta);
    end
    H_normalized = H ./ repmat(sqrt(sum(H.^2, 2)), 1, parameters.cluster_count);

    stream = RandStream.getGlobalStream;
    reset(stream);
    state.clustering = kmeans(H_normalized, parameters.cluster_count, 'MaxIter', 1000, 'Replicates', 10);
    state.objective = objective;
    state.parameters = parameters;
    state.Theta = Theta;
    state.time = toc;
end

function K_Theta = calculate_localized_kernel_theta(K, Theta)
    K_Theta = zeros(size(K(:, :, 1)));
    for m = 1:size(K, 3)
        K_Theta = K_Theta + (Theta(:, m) * Theta(:, m)') .* K(:, :, m);
    end
end

File: lmkkmeans_train.R
library(Rmosek)

lmkkmeans_train <- function(Km, parameters) {
  state <- list()
  state$time <- system.time({
    N <- dim(Km)[2]
    P <- dim(Km)[3]
    Theta <- matrix(1 / P, N, P)
    K_Theta <- matrix(0, nrow(Km), ncol(Km))
    for (m in 1:P) {
      K_Theta <- K_Theta + (Theta[,m,drop = FALSE] %*% t(Theta[,m,drop = FALSE])) * Km[,,m]  
    }

    objective <- rep(0, parameters$iteration_count)
    for (iter in 1:parameters$iteration_count) {
      print(sprintf("running iteration %d...", iter))
      H <- eigen(K_Theta, symmetric = TRUE)$vectors[, 1:parameters$cluster_count]
      HHT <- H %*% t(H)

      Q <- matrix(0, N * P, N * P)
      for (m in 1:P) {
        start_index <- (m - 1) * N + 1
        end_index <- m * N
        Q[start_index:end_index, start_index:end_index] <- diag(1, N, N) * Km[,,m] - HHT * Km[,,m]
      }
      problem <- list()
      problem$sense <- "min"
      problem$c <- rep(0, N * P)
      problem$A <- Matrix(rep(diag(1, N, N), P), nrow = N, ncol = N * P, sparse = TRUE)
      problem$bc <- rbind(blc = rep(1, N), buc = rep(1, N))
      problem$bx <- rbind(blx = rep(0, N * P), bux = rep(1, N * P))
      I <- matrix(1:(N * P), N * P, N * P, byrow = FALSE)
      J <- matrix(1:(N * P), N * P, N * P, byrow = TRUE)
      problem$qobj <- list(i = I[lower.tri(I, diag = TRUE)],
                           j = J[lower.tri(J, diag = TRUE)],
                           v = Q[lower.tri(Q, diag = TRUE)])
      opts <- list()
      opts$verbose <- 0
      result <- mosek(problem, opts)
      Theta <- matrix(result$sol$itr$xx, N, P, byrow = FALSE)
      K_Theta <- matrix(0, nrow(Km), ncol(Km))
      for (m in 1:P) {
        K_Theta <- K_Theta + (Theta[,m,drop = FALSE] %*% t(Theta[,m,drop = FALSE])) * Km[,,m]  
      }

      objective[iter] <- sum(diag(t(H) %*% K_Theta %*% H)) - sum(diag(K_Theta))
    }
    H_normalized <- H / matrix(sqrt(rowSums(H^2, 2)), nrow(H), parameters$cluster_count, byrow = FALSE)

    set.seed(NULL)
    state$clustering <- kmeans(H_normalized, centers = parameters$cluster_count, iter.max = 1000, nstart = 10)$cluster
    state$objective <- objective
    state$parameters <- parameters
    state$Theta <- Theta
  })
  return(state)
}
